src/LinuxTracing/PerfEventReaders.cpp

```diff
@@ -303,11 +303,11 @@ SchedWakeupPerfEvent ConsumeSchedWakeupPerfEvent(PerfEventRingBuffer* ring_buffe
       .timestamp = ring_buffer_record.sample_id.time,
       .ordered_stream = PerfEventOrderedStream::FileDescriptor(ring_buffer->GetFileDescriptor()),
       .data =
-          {
-              // The tracepoint format calls the woken tid "data.pid" but it's effectively the
-              // thread id.
-              .woken_tid = sched_wakeup.pid,
-          },
+          {// The tracepoint format calls the woken tid "data.pid" but it's effectively the
+           // thread id.
+           .woken_tid = sched_wakeup.pid,
+           .was_blocked_by_thread = static_cast<pid_t>(ring_buffer_record.sample_id.tid),
+           .was_blocked_by_process = static_cast<pid_t>(ring_buffer_record.sample_id.pid)},
```

Comment:
We have this issue with `sched:sched_switch` where in some cases `ring_buffer_record.sample_id.pid` is -1 even if `ring_buffer_record.sample_id.tid` is not (see the `SchedSwitchPerfEventData.prev_pid_or_minus_one` field). `SwitchesStatesNamesVisitor` then has logic to deal with it.

That's only when the thread is exiting, so it might never happen for this tracepoint, however I would ask you to verify that.

Reply:
Ok well, we have learned to take nothing for granted with `perf_event_open`, tracepoints, etc. 

Anyway, looking again at `SchedSwitchPerfEventData`, there we have two sources of the tid, `ring_buffer_record.sample_id.tid` (this could be -1 like `ring_buffer_record.sample_id.pid`) and `ring_buffer_record.data.prev_pid`, so we can deal with `ring_buffer_record.sample_id.pid == -1`. Here we seemingly only have one source for the tid, `ring_buffer_record.sample_id.tid`, so if this is -1 for whatever reason, there doesn't seem to be much we can do.

So in practice there's not much to do. However, my general point here and in other comments was that we need to be aware of what possible values are: Are these always regular pids/tids (>0)? Can they be 0? Can they be -1? If you are set up to run a game or a benchmark, a quick run with a bit of logging here could help understand that.